File: design_patterns/002_observer/solutions/python/weatherStation/src/WeatherData.py
from enum import Enum
from typing import List

from IObserver import IObserver
from IPublisher import IPublisher


class WeatherInfo(Enum):
    TEMPERATURE     = 'temperature'
    HUMIDITY        = 'humidity'
    PRESSURE        = 'pressure'


class WeatherData(IPublisher):
    _observers: List[IObserver]

    def __init__(self) -> None:
        self._observers = []
        super().__init__()


    def subscribe(self, observer: IObserver) -> None:
        self._observers.append(observer)


    def unsubscribe(self, observer:IObserver) -> None:
        print(f'---- Unsubscribing: {observer.__class__.__name__}')
        self._observers.remove(observer)


    def notify(self, data:dict) -> None:
        for observer in self._observers:
            observer.update(data)


    def set_measurements(self, temperature:float=None, humidity:float=None, pressure:float=None) -> None:
        updated_measurements:dict = {}
        updated_list = []
        if temperature is not None:
            updated_list.append(WeatherInfo.TEMPERATURE)
            updated_measurements[WeatherInfo.TEMPERATURE] = temperature
        if humidity is not None:
            updated_list.append(WeatherInfo.HUMIDITY)
            updated_measurements[WeatherInfo.HUMIDITY] = humidity
        if pressure is not None:
            updated_list.append(WeatherInfo.PRESSURE)
            updated_measurements[WeatherInfo.PRESSURE] = pressure
        print(f'Updating: {[updated.value for updated in updated_list]}')
        self.notify(updated_measurements)
        print('')


if __name__ == '__main__':
    raise Exception(f'This module should not be executed directly. Only for imports')

File: design_patterns/002_observer/solutions/typescript/weatherStation/src/WeatherData.ts
import { IObserver } from './IObserver';
import { IPublisher } from './IPublisher';

export enum WeatherInfo {
  TEMPERATURE = 'temperature',
  HUMIDITY = 'humidity',
  PRESSURE = 'pressure',
}

export type WeatherInfoType = {
  temperature?: number;
  humidity?: number;
  pressure?: number;
};

export class WeatherData implements IPublisher {
  private _observers: IObserver[];

  constructor() {
    this._observers = [];
  }

  subscribe(observer: IObserver): void {
    this._observers.push(observer);
  }

  unsubscribe(observer: IObserver): void {
    console.log(`---- Unsubscribing: ${observer.constructor.name}`);
    const index = this._observers.indexOf(observer);
    if (index !== -1) {
      this._observers.splice(index, 1);
    }
  }

  notify(data: { [key: string]: any }): void {
    for (const observer of this._observers) {
      observer.update(data);
    }
  }

  setMeasurements(data: WeatherInfoType): void {
    const updatedMeasurements: { [key: string]: any } = {};
    const updatedList: WeatherInfo[] = [];

    if (data.temperature !== undefined) {
      updatedList.push(WeatherInfo.TEMPERATURE);
      updatedMeasurements[WeatherInfo.TEMPERATURE] = data.temperature;
    }
    if (data.humidity !== undefined) {
      updatedList.push(WeatherInfo.HUMIDITY);
      updatedMeasurements[WeatherInfo.HUMIDITY] = data.humidity;
    }
    if (data.pressure !== undefined) {
      updatedList.push(WeatherInfo.PRESSURE);
      updatedMeasurements[WeatherInfo.PRESSURE] = data.pressure;
    }

    console.log(`Updating: ${updatedList.map(updated => updated)}\n`);
    this.notify(updatedMeasurements);
  }
}
